capstone/src/main/java/com/google/sps/servlets/ChallengeServlet.java

```diff
@@ -0,0 +1,64 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import java.sql.Timestamp;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import com.google.sps.Objects.Challenge;
+
+@WebServlet("challenge")
+public class ChallengeServlet extends HttpServlet {
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query("Challenge").addSort("dueDate", SortDirection.DESCENDING);
+    PreparedQuery results = datastore.prepare(query);
+    // Get most recent challenge in database
+    Entity entity = results.asIterable().iterator().next();
+    String challengeName = (String) entity.getProperty("name");
+    long dueDate = (long) entity.getProperty("dueDate");
+    ArrayList<String> usersCompleted = (ArrayList<String>) entity.getProperty("usersCompleted");
+    // setting badge as null for now
+    Challenge challenge = new Challenge(challengeName, dueDate, null, usersCompleted);
+    String json = new Gson().toJson(challenge);
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String challengeName = request.getParameter("name");
+    LocalDateTime dueDate = this.getDueDate(LocalDateTime.now());
+    long dueDateMillis = Timestamp.valueOf(dueDate).getTime();
+    Entity challengeEntity = new Entity("Challenge");
+    challengeEntity.setProperty("name", challengeName);
+    challengeEntity.setProperty("dueDate", dueDateMillis);
+    challengeEntity.setProperty("usersCompleted", new ArrayList<String>());
```

Comment:
consider doing this in the `Challenge.java` file

Solution 1:
```java
  Challenge challenge = new Challenge(data, that, i, need);
  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
  datastore.put(challenge.toEntity());
```

or even

Solution 2:
```java
  Challenge challenge = new Challenge(data, that, i, need);
  challenge.save();
```

I think I like the first one because then your classes don't have to be aware of the DB.

But if you wanted to integrate the DB into the class objects, I would recommend doing it with inheritance so you can avoid implementing a lot of duplicate logic.

I think if you do Solution 1, you should consider adding an Interface so all your objects have a `fromEntity(Entity entity)` and `toEntity()` function that they have to implement. Reach out if you have questions about interfaces/implementing this/etc.